### DISCUSSION TITLE
Общие замечания по реализациям

### DISCUSSION THREAD

**Discussion by GOOD-Stuff** [GOOD-Stuff/ssw_lab_4sem#16] [Show and tell]
Вижу, что вы начинаете потихоньку вносить свои реализации в репозиториях, и потому хотел бы заранее предостеречь от ошибок, к-ые нахожу в некоторых работах. Скорее всего все дальнейшие замечания так же буду добавлять сюда (надеюсь, про то как работают гиперссылки вы знаете):

1. [используйте](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rs-guards) _header guards_ вместо `#pragma once`. 
```C++
// в файле my_code.h
#ifndef MY_CODE_H_
#define MY_CODE_H_

// здесь тело заголовочного файла и прочее

#endif
```

2. объявления полей класса [лучше реализовывайте](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#nl16-use-a-conventional-class-member-declaration-order) в следующей последовательности: сначала все публичные поля, затем защищённые и лишь в конце приватные. Пример:
```C++
class ExampleContainer {
    public:       
        void SetSomething(int);
    private:
        int data {0}; // braced initialization
};
```

3. не используйте директиву `friend` для реализации перегрузки операторов. Т.к. все ваши операторы сравнения работают именно со своими же классами (т.е. вы будете сравнивать между собой объекты одного класса), то нет никакой необходимости в указании `friend` для предоставления доступа к приватным полям. 
```C++
template <class T>
class ExampleContainer {
    public:
        bool operator==(const ExampleContainer& rhs); // вдобавок, аргумент должен передаваться по константной ссылке
    private:
        T data {}; 
};
```

4. методы класса, к-ые не модифицируют его поля должны быть объявлены как `const`:
```C++
template <class T>
class ExampleContainer {
    public:
        int GetCount() const { // модификатор const "говорит" компилятору, что поля класса не изменятся при вызове метода
            return count;
        }
    private:
        int count {0}; 
};
```

5. не указывайте явно `this` в методах класса для обращения к полям/методам класса, по крайне мере без явной необходимости:
```C++
// ПЛОХОЙ пример. Так делать НЕ НАДО
class ExampleContainer {
    public:
        void func () {
            this->field = this->boo(this->foo);
            this->boo(this->boo(this->boo(this->foo))); 
        }
    private:
        int foo {123};
        int field {0};
   
        int boo(int);
};
```
Для обращения к полям класса внутри метода не нужно явно указывать указатель на объект (`this`). Он неявно добавляется компилятором. Явное указание же лишь усложняет чтение кода и добавляет лишнего визуального шума.

6. всегда [инициализируйте](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es20-always-initialize-an-object) поля класса внутри описания класса, при этом [используйте _braced initialization_](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es23-prefer-the--initializer-syntax).

**Comment by GOOD-Stuff** [comment]
В продолжение темы накопились ещё несколько новых пунктов:

7. не используйте включение всего множества стандартной библиотеки в заголовочных файлах ([`using namespace std`](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#sf7-dont-write-using-namespace-at-global-scope-in-a-header-file)).  
 Вместо этого либо включайте отдельные её компоненты (например `cout`), либо сокращайте вложенностью в какую-то область видимости (скоуп), старайтесь избегать использовать включение глобально даже в файлах реализации (*.cpp*):
```C++
// header.h
#include <iostream>

// включаю в хедере возможность использовать стандартный вывод в консоль
using std::cout; // для общих сообщений
using std::cerr;  // для ошибок
using std::endl; // для переноса строки и очистки буфера

// хочу использовать зачем то в хедере вывод в консоль, например в шаблоне
template <class T>
T func(const T& obj) {
  // могу в хедере больше не указывать std::cout
  cout << "Some message" << endl;
}

// another.cpp
#include <chrono>

void func() {
  using namespace std::chrono;
  auto start = high_resolution_clock::now();
 
  // some calcs
  
  auto end = high_resolution_clock::now();
    
  auto int_ms = duration_cast<milliseconds>(end - start).count();
}
```  

8. [не выделяйте память для 0 байт](https://stackoverflow.com/questions/1087042/c-new-int0-will-it-allocate-memory). 
По стандарту вам конечно же могут аллоцировать (выделить) память (хотя `new` и в таких случаях может кинуть исключение), но такой указатель нельзя будет разыменовывать - это является *undefined behavior* (*UB*). 
Т.е. с таким указателем нельзя работать, вам в любом случае придётся его освободить и выделить память заново уже под нормальный размер. 

9. если у вас есть конструктор принимающий всего один аргумент, то его нужно указывать как [*explicit*](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-explicit). 
Добавление ключевого слова `explicit` позволит вам избежать ошибок неявного преобразования типа:
```C++
// ПЛОХОЙ пример:
class String {
public:
    String(int);   // вызов такого конструктора "не строгий" и может произойти под неявным преобразованием
    // ...
};

String s = 10;   // сконструирует объект ни смотря на наличие неявного преобразования
/**************************************************/
class String {
public:
    explicit String(int);   // "строгий" вызов конструктора, вы должны явно указать String(10), чтобы компилятор знал, что вы понимаете, что делаете
    // ...
};

String s = 10;   // ошибка: попытка конвертировать из типа int в не скалярный тип String
String s2 = String(10); // ошибки нет, вызовется конструктор String(int)
```
т.е. компилятор заранее вас убережёт от ошибок неявного преобразования, которое в более сложных примерах может сыграть плачевную роль;

10. для вычисления времени используйте [`std::chrono`](https://docs.w3cub.com/cpp/header/chrono), вместо устаревшей [`ctime.h`](https://docs.w3cub.com/cpp/header/ctime):
```C++
#include <chrono> 
using ms = std::chrono::milliseconds; 

void ex() { 
  auto start = std::chrono::high_resolution_clock::now(); 
// some computation 
  auto end = std::chrono::high_resolution_clock::now(); 
  auto res_time = std::chrono::duration_cast<ms>(end - start).count(); 
}
```

11. пожалуй одна из самых болезненных (для меня) ваших ошибок - некорректное применение директивы препроцессора `#include`. 
**[НЕ ИСПОЛЬЗУЙТЕ `include` С ФАЙЛАМИ РЕАЛИЗАЦИИ](https://isocpp.org/wiki/faq/templates#templates-defn-vs-decl)**!
т.е. **НЕ пишите** такое:
```C++
#include <my_template.cpp>
#include <vector.cpp>
#include <why?!.cpp> 
```
Если же вы реализовываете шаблон, то у вас есть [всего пару](https://isocpp.org/wiki/faq/templates#separate-template-fn-defn-from-decl) способов собрать корректно:
- вся реализация, как и декларация шаблона будет в хедере (в заголовочном файле - *.h*);
- декларация в хедере, реализация в *.cpp*, но в конец *.cpp* придётся добавить инстанцирование шаблона тех типов, для которых он будет использоваться, т.е.:
```C++
// в foo.cpp
#include <foo.h> // здесь хранится объявление шаблона класса foo

template<class T>
void foo<T>::some_func() {
 // какая то реализация
}

// ещё всякий код
template class foo<int>; // явное инстанцирование шаблона для типа int
```
- декларация в хедере, реализация в отдельном *.cpp* (лучше в таком случае указывать тип *.tpp*), в конец хедера инклудится реализация:
```C++
// в foo.h
#ifndef FOO_H_ // не забываем про хедер гуарды, иначе у вас будут проблемы с компиляцией
#define FOO_H_ 

template <class T>
class foo {
 // объявление методов/полей
};

#include <foo.tpp> // директива #include просто копирует содержимое файла, поэтому тут будет вся реализация

#endif

// в foo.tpp
#include "foo.h"

template <class T>
// реализация методов шаблонного класса
```
Проблема в том, что шаблоны компилируются несколько иначе, нежели ваш обычный код. 
Компилятор должен ещё на этапе сборки знать типы которыми инстанцируется (создаётся) ваш шаблонный класс/функция. 
Т.е. у вас в коде где то явно должен быть указан вызов вашего шаблона с конкретным типом, при этом компилятор должен ещё знать о реализации функций вашего шаблона, следовательно реализация шаблона должна быть где-то поблизости с вызовом объекта-шаблона. 
На практике первый вариант (реализация в *.h*) более предпочтителен, по крайне мере если сам шаблон не особо большой и имеет несложный функционал;

12. если реализуете конструктор копирования, то в таком случае реализуйте и оператор копирования (тоже самое и в обратном случае). При этом [**НЕ НАДО**](https://stackoverflow.com/questions/2639017/calling-assignment-operator-in-copy-constructor) в конструкторе копирования просто вызывать оператор копирования, а-ля: 
```C++
class foo {
  foo(const foo& obj) {
    *this = obj; // НЕ ДЕЛАЙТЕ ТАК!
  }
};
```
[Конструктор](https://www.cplusplus.com/articles/y8hv0pDG/) в первую очередь сначала **должен проинициализировать** объект, т.е. сконструировать поля класса. При этом поля класса могут иметь более сложный тип, чем классический `int` и `int*`. Эти типы тоже могут иметь специфический конструктор и прочие особенности (+ могут кидать исключения в конструкторе). 
Поэтому вы должны явно сначала проинициализировать поля и только [потом](http://courses.washington.edu/css342/zander/css332/copyeq.html) производить копирование.

13. в операторе копирования всегда [производите проверку](https://stackoverflow.com/questions/2639017/calling-assignment-operator-in-copy-constructor) на попытку копирования самого себя, либо используйте [*copy & swap*](https://stackoverflow.com/questions/3279543/what-is-the-copy-and-swap-idiom) идиому:
```C++
class foo {
  int* ptr {nullptr};
  public:
  // здесь всякие конструкторы/деструкторы, их реализацию я опущу
  foo& operator=(const foo& obj) {
    if (*this != obj) { // исходим из предположения, что попытка присвоить самого себя, крайне редка
      // производим здесь копирование 
    }
    return *this;
  }

  // либо идиома copy & swap
  foo& operator= (const foo& obj) {
    foo tmp = obj;
    std::swap(*this, tmp); // нет необходимости в проверке, просто обмениваемся полями с временным объектом
    return *this;
    
  }
};
```

**Reply by Overfirst** [reply]
`*this = obj; // НЕ ДЕЛАЙТЕ ТАК!` – если реализованы корректные операторы копирования и переноса, которые делают в точности то же, что должны делать соответствующие конструкторы, почему нельзя это использовать? Что касается инициализации специфических полей, то да – справедливо, но в данных ЛР (пока первых двух) такой проблемы не обнаружил.

upd: по поводу инициализации и последующей перезаписи – практически ничего не теряем (в данных ЛР).

`if (*this != obj)` – похоже на _**stack overflow exception**_)
Почему не `if (this != &obj)`?

**Reply by GOOD-Stuff** [reply]
>  если реализованы корректные операторы ...  которые делают в точности то же, что должны делать соответствующие конструкторы

Только проблема в том, что задачи у **операторов** и **конструкторов** разные и вызываться они должны лишь в соответствующих им операциях/ситуациях. 
При вызове оператора копирования из конструктора, вы пытаетесь делать копии в поля ещё не созданного объекта, к-ый только начинает свою инициализацию. 
Вдобавок, что произойдёт (при реализации через оператор), если конструктору копирования подать тот же самый объект или если у класса есть константные поля? Это либо потенциальное *UB*, либо вы просто получите некорректно проинициализированный объект (не говоря уже о случаях с применением наследований и т.п.).
Я уверен, что можно подобрать ещё несколько примеров, при которых реализация конструктора через оператор не будет работать правильно. 
Не всегда сокращения кода на пару-тройку строк качественно сказывается на всей кодовой базе и производительности работы. 

> но в данных ЛР (пока первых двух) такой проблемы не обнаружил

Это только потому что вы свои контейнеры используете с простыми типами, но если я проверки ради попробую использовать более сложные примеры, ваши работы могут дать не тот результат, к-ый вы от них ожидали. 
Мне важно, чтобы вы делали всё идиоматически корректно. 

> Почему не if (this != &obj)?

Как минимум потому что оператор `&` может быть перегружен и может не возвращать адрес объекта (или делать какой-нибудь нежелательный в такой ситуации сайд эффект). В таком уж случае хотя бы используйте вместо & вызов `std::addressof(obj)`. 
Но по большей части тут причина скорее стилистическая, т.к. раньше такой паттерн был более в ходу, вдобавок разыменовать `this` проще, чем корректно взять адрес от объекта, но это уже скорее ИМХО.
